CMakeLists.txt

```diff
@@ -1,538 +1,396 @@
+  message(STATUS "Using Matlab pre-2018a API for mxGetPr compatibility")
+
+  if(BUILD_TESTING)
+    message(
+      STATUS "Disabling BUILD_TESTING because we are building Matlab interface")
+    set(BUILD_TESTING OFF)
+  endif(BUILD_TESTING)
+endif(MATLAB)
+
+if(R_LANG)
+  find_package(R)
+  if(NOT R_FOUND)
+    message(FATAL_ERROR "You need R libraries to build the R interface")
+  endif(NOT R_FOUND)
+
+  message(STATUS "R exec is: " ${R_EXEC})
+  message(STATUS "R root dir is: " ${R_ROOT_DIR})
+  message(STATUS "R includes are in: " ${R_INCLUDE_DIRS})
+
+  include_directories(${R_INCLUDE_DIRS})
+  add_definitions(-DR_LANG)
+
+  if(BUILD_TESTING)
+    message(
+      STATUS "Disabling UNITTESTS because we are building the R interface")
+    set(BUILD_TESTING OFF)
+  endif(BUILD_TESTING)
+endif(R_LANG)
 
-include(GNUInstallDirs)
-
-install(TARGETS osqpstatic
-        EXPORT  ${PROJECT_NAME}
-        ARCHIVE       DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-        LIBRARY       DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-        RUNTIME       DESTINATION "${CMAKE_INSTALL_BINDIR}")
-
-
-# Install Headers
 # ----------------------------------------------
+# Static Library - osqpstatic
+# ----------------------------------------------
+function(OSQP_BUILD_STATIC_LIB)
+  add_library(osqpstatic STATIC $<TARGET_OBJECTS:OSQPLIB>)
+  if(ALGEBRA_DEFAULT)
+    # Transitive dependency on OBJECT library does not work. See
+    # https://gitlab.kitware.com/cmake/cmake/-/issues/18682
+    target_sources(osqpstatic PRIVATE $<TARGET_OBJECTS:qdldlobject>)
+  endif()
+
+  if(MATLAB)
+    target_link_libraries(osqpstatic ${Matlab_LIBRARIES})
+  endif(MATLAB)
+endfunction()
 
-install(FILES ${osqp_headers} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/osqp")
+# ----------------------------------------------
+# Shared Library - osqp
+# ----------------------------------------------
+function(OSQP_BUILD_SHARED_LIB)
+  add_library(osqp SHARED $<TARGET_OBJECTS:OSQPLIB>)
+  target_include_directories(
+    osqp
+    PUBLIC
+      "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+      "$<INSTALL_INTERFACE:$<INSTALL_PREFIX>/${CMAKE_INSTALL_INCLUDEDIR}/osqp>")
+endfunction()
 
+# ----------------------------------------------
+# Application - osqp_demo
+# ----------------------------------------------
+function(OSQP_BUILD_DEMO)
+  add_executable(osqp_demo ${PROJECT_SOURCE_DIR}/examples/osqp_demo.c
+                           ${osqplib_sources})
+  target_include_directories(osqp_demo PRIVATE ${osqplib_includes})
+  if(ALGEBRA_MKL)
+    target_link_libraries(osqp_demo osqpstatic $<LINK_ONLY:MKL::MKL>)
+  elseif(ALGEBRA_CUDA)
+    target_link_libraries(osqp_demo osqpstatic cublas cusparse)
+  endif()
+endfunction()
 
-if (CUDA_SUPPORT)
-    target_link_libraries (osqpstatic cublas cusparse)
+# ----------------------------------------------
+if(NOT PYTHON
+   AND NOT MATLAB
+   AND NOT R_LANG
+   AND NOT EMBEDDED)
+  osqp_build_static_lib()
+  osqp_build_shared_lib()
```

Comment:
We are building `osqp` libraries under the conditions above, but are trying to install them below without imposing the same conditions.

Reply:
fixed